### DISCUSSION TITLE
State of React Native 2022 Call for Questions

### DISCUSSION THREAD

**Discussion by bbarthec** [software-mansion/state-of-react-native#1] [Preview / Call for Questions]
## Call for Questions 2022 is now closed ⛔ 

Below we present areas that we plan to ask about in the survey. If you have any suggestions on how to improve any of these, let us know here.

We'll review all provided suggestions and try our best to select these that fit into our vision of the survey.

### Platforms API

Section asking about familiarity with variety of APIs available on each platform that React Native supports. We've prepared a rich list of these already internally at @software-mansion, but we would love to get some feedback from the community as well. If there are APIs that need to appear in this survey, be sure to inform us!

- e.g. Camera, SQL databases, NFC

### React Native app development aspects

Section asking about available solutions to following aspects of app development. For each aspect we plan to list commonly used techniques/libraries. If there are solutions you wish should appear in each section or there are sections that are missing from our list, let us know! Below we present you a few ideas for each category.

- State Management
  - e.g. Redux, Zustand
- Data Fetching
  - e.g. fetch API, relay
- Styling 
  - e.g. StyleSheet API, NativeWind
- Component libraries
  - e.g. React Native Paper, ui-kitten
- Navigation
  - e.g. react-navigation, expo-router, react-native-navigation
- Animations
  - e.g. Animated API, Moti, reanimated, Skia
- Gestures
  - e.g. PanResponder API, react-native-gesture-handler
- Crash Reporting
  - e.g. Sentry, Datadog
- Storage
  - e.g. AsyncStorage, react-native-mmkv
- OTA updates
  - e.g. EAS Update, CodePush 

### React Native features

In this section we plan to ask about awareness of mechanics that are specific to React Native technology and are almost always not available in other technologies. If there is any React Native aspect that you consider crucial, let us know!

- e.g. Fabric, JSI, turbo modules

### Developer tools

This section is all about your workspace. Share with us the most important tooling you're using working with React Native projects on a daily basis!

- e.g. Storybook, GitHub Copilot, volta

### Opinions about React Native ecosystem

Section about how you feel using React Native. We plan to highlight the most prominent aspects of this technology as well as the most annoying ones.

- e.g. 
  - 👍 flex gap coming with 0.71
  - 👎 no other bundler than Metro available
  - 👍 good cooperation between Meta & Community via [react-native-community/discussions-and-proposals](https://github.com/react-native-community/discussions-and-proposals/discussions) and [reactwg/react-native-releases](https://github.com/reactwg/react-native-releases/discussions)

**Comment by marklawlor** [comment]
Some suggestions

## React Native app development aspects

- authentication
- testing
- accessibility
- data visualization

## Resources

- Learning methods
- Podcasts
- Content creators
- Communities (discord/slack channels)

## Ecosystem

- Do you develop for one or multiple platforms
- Does your app have another version written in a different framework (eg a web version written in React DOM)
- Awareness of off-tree platforms (web/windows/osx/tv/watch)
- Starter templates (create-expo-app, ignite, solito, etc)
- Monorepo tools

**Reply by bbarthec** [reply]
@marklawlor, thank you for all these suggestions 👏 Most of them landed directly or in a converted form 😁

**Comment by julienqueffelec** [comment]
hey,

## React Native app development aspects
- Push Notifications (Firebase, OneSignal...)

## Developer tools
- E2E tests (Appiuem, Detox, Maestro)
- Automation tools (Fastlane, EAS)
- Debugger (React Native Debugger, Flipper)

## Opinions about React Native ecosystem
- Do you plan to use react native web in a future project?
- bare react-native vs expo managed workflow usage

**Reply by bbarthec** [reply]
@julienqueffelec, thank you for all these suggestions 👏 All of these were very inspiring and most of them landed directly or in a converted form 😁

**Comment by slorber** [comment]
Cross-platform abstractions: React-Native-Web, Solito, Moti, Drispy, Tamagui, StyleX, `.web.js`, React-Three-Fiber RN...

Newsletters: This Week In React, React-Native Newsletter, React-Native Now, React Newsletter, React Status...

**Reply by bbarthec** [reply]
@slorber, thank you for all these suggestions 👏 Around half of them landed directly or in a converted form 😁

**Comment by mateusz1913** [comment]
👋 

## React Native app development aspects
- Keyboard handling (KeyboardAvoidingView, react-native-keyboard-aware-scroll-view, react-native-keyboard-manager, react-native-keyboard-controller, react-native-avoid-softinput)

## Developer tools
- Git hooks (Husky, Lefthook, etc.)
- Linting tools (eslint, prettier, lint-staged, swiftlint, clang-format, spotless, ktlint, etc.)

## Opinions about React Native ecosystem
Do you prefer to use 3rd party libraries with native+JS or JS-only implementation?
Do you prefer to upgrade your app manually with upgrade-helper or with CLI (React Native or Expo)? How would you rate upgrade experience with CLI?

**Reply by bbarthec** [reply]
@mateusz1913, thank you for all these suggestions 👏 These ideas were very inspiring, but some of them are too niche in our opinion and some other are duplicating the work done by State of JS survey 😉 Nevertheless, the survey allows for answering some of them indirectly 😁

**Comment by jadonshubham** [comment]
### Performance Aspects of React Native Apps

- Experience of making apps run at 60fps for low end android devices.
- Does performance bottleneck coming in way of making something good? (like dropping some crazy feature so, everything is smooth in app).
- Do you feel new architecture will solve all these performance problems?

**Reply by bbarthec** [reply]
@jadonshubham, thank you for all these suggestions 👏 These could lit a very inspiring discussions among devs, but we think that these topics are very broad and we failed to convert them into questions fitting our vision of the survey 😓 Nevertheless, the survey allows for answering some of them indirectly 😁

**Comment by Foosballfan** [comment]
### What UI/End-to-End testing do you do in your React Native app?
- Automated with Detox
- Automated with native testing frameworks, e.g. XCUITest and Espresso
- Automated with Appium
- Manual
- None
- Other, please specify

**Reply by bbarthec** [reply]
@Foosballfan, thank you for this suggestion 👏 We've managed to include this in the survey 😁

**Comment by travisobregon** [comment]
I think seeing [Legend State](https://www.legendapp.com/open-source/state/) for "State Management" and [Legend Motion](https://www.legendapp.com/open-source/motion/) for "Animations" would be good.

**Reply by bbarthec** [reply]
@travisobregon, thank you for this suggestion 👏 We've managed to include this in the survey 😁

**Comment by usrbowe** [comment]
### 🧰   Developer tools: Flipper  🟣

#### Official plugins
It would be good to push for introducing "official" plugins for major community packages for Flipper. Which are part of the package and well maintained with feature parity. Current state of plugins is quite confusing for newcomers as it's unclear which one to choose.

**Reply by marklawlor** [reply]
This is a call for questions for a community survey, not questions to the RN core team. Could you re-phase your points so they are questions that would be appropriate for the community? Or maybe this feedback would be more appropriate in the [2022: How can we improve React Native](https://github.com/react-native-community/discussions-and-proposals/discussions/528) discussion

**Reply by usrbowe** [reply]
oh good point, I've updated the topic to focus on community only.

**Reply by bbarthec** [reply]
@usrbowe, thank you for this suggestion 👏 We've failed to include this question in the survey 😓 Nevertheless, the survey allows for answering this question indirectly 😉

**Comment by bbarthec** [comment]
I'm glad to announce that we're almost there! 🚀😊

The preview version of the survey is available here: https://survey.stateofreactnative.com - we invite you to check it out 😁 
We plan to iterate over it a little more (mostly fixing bugs & wording, reducing loading times - we know that for some users first load takes far toooooo long) and release the final version by the end of this week (yes, before Christmas 🎄) 🔥

I'd like to say big thank you to everyone that contributed towards the survey! 👏👏👏

**Comment by bbarthec** [comment]
I'm closing this discussion and invite you all to use followings:
- https://github.com/software-mansion/state-of-react-native/issues/3 for content discussion
- https://github.com/software-mansion/state-of-react-native/issues/2 for reporting technical issues
